ksonnet-gen/ksonnet/emit.go

```diff
@@ -422,6 +429,20 @@ func (ao *apiObject) emitAsRefMixins(
 	m.writeLine("},")
 }
 
+func (ao *apiObject) emitConstructor(m *indentWriter) {
+	if dm, ok := ao.propertyMethods["default"]; ok {
+		log.Fatalf(
+			"Attempted to create constructor, but 'default' property already existed at '%s'",
+			dm.path)
+	}
+
+	if _, ok := ao.propertyMethods["metadata"]; !ok {
+		m.writeLine("default():: apiVersion + kind + {},")
+	} else {
+		m.writeLine("default(name):: apiVersion + kind + self.mixin.metadata.name(name),")
```

Comment:
Good to know about `generateName`. I had no idea. That's a good enough reason to not take `name` as a default parameter.

But, just btw, regarding the second point, we're not blacking out all `metadata` mixins, we're only blacking out `metadata` mixins when they are referencing `ListMeta` (as opposed to `ObjectMeta`). So the effect is that anything with a `metadata` property that references `ObjectMeta` will end up getting a `name` property. As I said, given `generateName`, we still have to get rid of name, but I just wanted to be clear about what this is generating.